### PR TITLE
[Translation] add fix to ignore empty translation

### PR DESCRIPTION
**Pull request by hansnilsson**
| Q | A |
| --- | --- |
| Bug fix? | [yes] |
| New feature? | [no] |
| BC breaks? | [unsure] |
| Deprecations? | [no] |
| Tests pass? | [yes] |
| Fixed tickets | #13483 |
| License | MIT |
| Doc PR | [] |

The expected behaviour is to treat empty translations like PHP gettext do, and show the fallback or original message instead.

### CODE REVIEW

File: src/Symfony/Component/Translation/MessageCatalogue.php [hunk 1]
```diff
@@ -120,7 +120,7 @@ public function defines($id, $domain = 'messages')
      */
     public function get($id, $domain = 'messages')
     {
-        if (isset($this->messages[$domain][$id])) {
+        if (isset($this->messages[$domain][$id]) && !empty($this->messages[$domain][$id])) {
```
Comment (inso) [inline src/Symfony/Component/Translation/MessageCatalogue.php]: should be `'' !== $this->messages[$domain][$id]` because `empty('0') === true`

**Comment by hansnilsson** [comment]
Thanks @inso - updated.

**Comment by timglabisch** [comment]
:-1: an empty translation can be helpful and it's a BC

**Comment by ogizanagi** [comment]
@timglabisch : How an empty translation could be helpful ? Do you have an example ?

**Comment by hansnilsson** [comment]
I can't see how empty slots in the UI can be helpful, I'd rather have the original string presented, and I'd prefer to have it behave like php gettext.

It's not hard to switch the Translator class though and override the behaviour there, so I won't be sorry if this is denied.

**Comment by Triiistan** [comment]
@ogizanagi https://github.com/symfony/symfony/issues/13483#issuecomment-71172886 is an example where it can be useful. So perhaps it needs more discussion to ensure we want to remove this particular side-effect.

**Comment by fabpot** [comment]
I'm :+1: on this change, which should be merged into master as this is a BC break. This BC break should be documented in the CHANGELOG for Symfony 2.7 as well.

ping @symfony/deciders

**Comment by weaverryan** [comment]
Hmm, I'm :-1: on this - it seems kind of like WTF behavior. I can imagine someone trying to *not* display something in a language (like @Triiistan is describing) and it seeming to not work - I set the key to `''`, but the key (or fallback) keeps dumping out. Why isn't my translation working!? :). 

I see that this was proposed to be consistent with `gettext`, and that definitely gives it weight. But I feel I'd want it configurable, if that's possible.

**Comment by fabpot** [comment]
Making such a behavior configurable is not an option and would be even worse (keep in mind that such configurability would impact third party translations as well).

Using empty translations to not display something is a hack, and we should not support such hacks.

What about changing the current behavior in 3.0?

**Comment by weaverryan** [comment]
Good point on 3rd party translations :).

But it reminds me of how I use Twig blocks - where I purposefully set blocks to a blank string (or null) to make them disappear. If you *were* to want to hide something for a locale, I guess you should use an empty space? I'm sure this is an uncommon situation, so I don't want to derail things for it. So I'll remove my -1 at least - I'm curious what others think, and I *really do* like that this is consistent with gettext.

**Comment by hansnilsson** [comment]
I've added a 3.0 upgrade message as well. Should this PR be closed and a new one opened for master, if the decision is to merge?

**Comment by stof** [comment]
Note that if you use Yaml for your translations, you can set the value to ``null`` for it to be ignored currently (while an empty string will be considered as a translation giving an empty string). I'm not sure other formats have a way to mark translations as ``null``

**Comment by Rvanlaak** [comment]
We're using Yaml and are using `""` as a translation value for languages where we don't want to show an element, but this isn't `null` right?
```yaml
translation_page:
    translation_key: ""
```
Switching based on the request locale in Twig is the preferred way, but in our project the developers are the ones managing the Twig templates and the translators (for now) need an easy way to hide some items.

:+1:  for 2.7 if the `is_null` check will be chosen

**Comment by stof** [comment]
@Rvanlaak ``null`` is already handled this way because of the ``isset`` check. There is no change needed

**Comment by fabpot** [comment]
Thinking about this a bit more, I don't think that using an empty string to make things disappear is really a good use case; I mean, most of the time, the string is surrounded by some HTML tags as well, which should be removed as well. I'm still thinking that this is a hack, and an unsupported one. We can of course merge this into the master branch if we think many people rely on the current behavior.

**Comment by hansnilsson** [comment]
Let me know if I need to add anything to the PR to make it mergable.

**Comment by Rvanlaak** [comment]
@fabpot this is also why I'd need the `translator.fallback` config to be "undefined" because the surrounding HTML is hidden with the following ugly code:
```twig
{% if ('order2.tooltip_'~service.id)|trans != ('order2.tooltip_'~service.id) %}
```
But it seems this is still possible after this PR, when fallback is set to 'undefined'.

Having a check like `hasTranslation` for the active locale, or another easier way for not having to check everything in the Controller would be usefull. That way I can get rid of such hacks. What about you @weaverryan ?

**Comment by fabpot** [comment]
ping @symfony/deciders We need to make a decision about this one. I'm :+1: for the record.

**Comment by xabbuh** [comment]
I am :+1: to add it, but I think it should only be done in 3.0 (I am not sure how many people rely on the current behaviour).

**Comment by Tobion** [comment]
I dort See why why we should create a bc break and remove a use case(empty translation) for noch real gain. The current Behälter majestätischen more sense to me Since IT also Covers both ways. So im Rächer -1

**Comment by Tobion** [comment]
Stupid auto correction and mobil Version cannot Edit

**Comment by stof** [comment]
@Tobion I tried to guess the original message. Tell me if I edited it the wrong way :smile:

**Comment by fabpot** [comment]
@stof I loved the original version of @Tobion message more... especially because we are talking about translation.

**Comment by stof** [comment]
@fabpot fair enough :smile:

**Comment by me2resh** [comment]
@fabpot 
I see this pull request still not merged, should I recreate one, I am having the same issue in https://github.com/symfony/symfony/issues/13483

**Comment by sstok** [comment]
@me2resh This pull-request was closed and therefor cannot be merged, doing this would be a BC break. For PHP/YAML we could decide to consider null (not empty) as undefined and make it fallback, for xliff however I don't know what could work.

**Comment by ghisleouf** [comment]
Hi guys,
What about reconsidering this PR ?

Actually, working with xliff files and translation tools provide you generated xliff files with empty `<target>` tags.
Without this test, fallback is not working anymore.

Thanks for your feedback.

**Comment by ywarnier** [comment]
Hi @fabpot @symfony/deciders, I would like to ask for this PR to be reconsidered (albeit on a newer branch). Here is my argument.

In Gettext's documentation, it is made clear that the default behaviour for untranslated strings is to be empty: https://www.gnu.org/software/gettext/manual/gettext.html#Untranslated-Entries

> When xgettext originally creates a PO file, unless told otherwise, it initializes the msgid field with the untranslated string, and leaves the msgstr string to be empty. Such entries, having an empty translation, are said to be untranslated entries.

The current check in `MessageCatalogue::defines()` only checks if the string is set, which returns `true` for empty strings.
This prevents the fallback mechanism to work, as it only kicks in when the string is not set.

Weblate, an Open Source app to deal with (software) translations respects the Gettext default and returns .po files with `msgstr ""` if no translation exists yet for a specific language. There is no way, currently, to configure or hack this without spawning a fork of Translation, except maybe to deal with the whole fallback mechanism before calling `trans()`, which feels like an absolute bummer when there is a fallback mechanism that is supposed to already "just work".

I'm not sure why @fabpot stated that:
> Making such a behavior configurable is not an option and would be even worse (keep in mind that such configurability would impact third party translations as well).

I don't see why having the (non-default) *option* to ask for a stricter check would be damaging to third party translations. Based on the default behaviour of Gettext, it seems more like offering a way to actually support the standard.

**Comment by nicolas-grekas** [comment]
Note that comments on closed issues/PRs are likely going to be lost.

**Comment by ywarnier** [comment]
@nicolas-grekas Should I open a new PR? I don't mind. Just want to maximize chances of it being well received :-)

**Pull request closed without merging**